js/src/forum/state/DiscussionListState.js

```diff
@@ -0,0 +1,153 @@
+import Discussion from '../../common/models/Discussion';
+
+export default class DiscussionListState {
+  constructor({ params = {} } = {}) {
+    this.params = params;
+
+    this.discussions = [];
+
+    this.moreResults = false;
+
+    this.loading = false;
+  }
+
+  /**
+   * Get the parameters that should be passed in the API request to get
+   * discussion results.
+   *
+   * @api
+   */
+  requestParams() {
+    const params = { include: ['user', 'lastPostedUser'], filter: {} };
+
+    params.sort = this.sortMap()[this.params.sort];
+
+    if (this.params.q) {
+      params.filter.q = this.params.q;
+
+      params.include.push('mostRelevantPost', 'mostRelevantPost.user');
+    }
+
+    return params;
+  }
+
+  /**
+   * Get a map of sort keys (which appear in the URL, and are used for
+   * translation) to the API sort value that they represent.
+   */
+  sortMap() {
+    const map = {};
+
+    if (this.params.q) {
+      map.relevance = '';
+    }
+    map.latest = '-lastPostedAt';
+    map.top = '-commentCount';
+    map.newest = '-createdAt';
+    map.oldest = 'createdAt';
+
+    return map;
+  }
+
+  /**
+   * Set the search parameters.
+   */
+  setParams(params) {
+    this.params = params;
+  }
+
+  /**
+   * Clear cached discussions.
+   */
+  clear() {
+    this.discussions = [];
+  }
+
+  /**
+   * Clear and reload the discussion list.
+   */
+  refresh(clear = true) {
+    if (clear) {
+      this.loading = true;
+      this.clear();
+    }
+
+    return this.loadResults().then(
+      (results) => {
+        this.discussions = [];
+        this.parseResults(results);
+      },
+      () => {
+        this.loading = false;
+        m.redraw();
+      }
+    );
+  }
+
+  /**
+   * Load a new page of discussion results.
+   *
+   * @param offset The index to start the page at.
+   */
+  loadResults(offset) {
+    const preloadedDiscussions = app.preloadedApiDocument();
+
+    if (preloadedDiscussions) {
+      return Promise.resolve(preloadedDiscussions);
+    }
+
+    const params = this.requestParams();
+    params.page = { offset };
+    params.include = params.include.join(',');
+
+    return app.store.find('discussions', params);
```

Comment:
Do we want to pass in the `app` (or store, or whatever dependencies there are) to the state as well? :thinking: That would make it easier to test them, potentially... :thinking:

Reply:
Isn't that what's done here? App is globally available

Reply:
That's the opposite of passing it in. We're accessing a global variable, I was suggesting to do dependency injection.

Reply:
Redux, Meiosis, and most other state management methodologies are based around having a single, global state object. Now, typically the main reason for wanting to avoid globals is that anything, anywhere can modify them, leading to unintended consequences. That's where the main point of these methodologies comes in: the modification of the global state is controlled by the methodology. For instance, with Meiosis, http://meiosis.js.org/tutorial/01-introduction.html

Reply:
I understand that concept, and the states and stores being a central place makes sense. I was discussing the method of accessing this central place: by calling on a global name (`app.store`) or by passing it in to the constructor (dependency injection). The latter not only makes tests easier, but it also allows passing in a different object with the same interface, should the need ever arise (and part of our extensibility philosophy is not knowing what use-cases might come up :laughing: ).

Reply:
Alright, I see your point here. I'll change it so that if `app` is needed by the states, it gets passed in on initialization (or should `app.store` be passed in?)

Reply:
After some more thought, I'm going to side with passing in the entire app. Pretty much every piece of the app could be reused in some way or another.

Reply:
Okay, thanks. No strong opinion there right now. We can revisit this when we have states everywhere, and see how it looks then / what patterns emerge / whether we want to refactor towards another solution.

Reply:
I definitely like where this is going already.